app/assets/javascripts/discourse/app/components/sidebar/edit-navigation-menu/categories-modal.gjs

```diff
@@ -18,126 +18,142 @@ import gt from "truth-helpers/helpers/gt";
     super(...arguments);
 
-    let categories = [...this.site.categories];
+    this.processing = false;
+    this.setFilterAndMode("", "everything");
+  }
+
+  setFilteredCategories(categories) {
+    const ancestors = findAncestors(categories);
+    const allCategories = categories.concat(ancestors).uniqBy((c) => c.id);
 
-    if (!this.siteSettings.fixed_category_positions) {
-      categories.sort((a, b) => a.name.localeCompare(b.name));
+    if (this.siteSettings.fixed_category_positions) {
+      allCategories.sort((a, b) => a.position - b.position);
+    } else {
+      allCategories.sort((a, b) => a.name.localeCompare(b.name));
     }
 
-    Category.sortCategories(categories).reduce(
-      (categoryGrouping, category, index, arr) => {
-        if (category.isUncategorizedCategory) {
-          return categoryGrouping;
-        }
+    this.filteredCategoriesGroupings = splitWhere(
+      Category.sortCategories(allCategories),
+      (category) => category.parent_category_id === undefined
+    );
 
-        categoryGrouping.push(category);
+    this.filteredCategoryIds = categories.map((c) => c.id);
+  }
 
-        const nextCategory = arr[index + 1];
+  async searchCategories(filter, mode) {
+    if (filter === "" && mode === "only-selected") {
+      this.setFilteredCategories(
+        await Category.asyncFindByIds(this.selectedSidebarCategoryIds)
+      );
+    } else {
+      const { categories } = await Category.asyncSearch(filter, {
+        includeAncestors: true,
+        includeUncategorized: false,
+      });
 
-        if (!nextCategory || nextCategory.level === 0) {
-          this.categoryGroupings.push(categoryGrouping);
-          return [];
+      const filteredFetchedCategories = categories.filter((c) => {
+        switch (mode) {
+          case "everything":
+            return true;
+          case "only-selected":
+            return this.selectedSidebarCategoryIds.includes(c.id);
+          case "only-unselected":
+            return !this.selectedSidebarCategoryIds.includes(c.id);
         }
+      });
 
-        return categoryGrouping;
-      },
-      []
-    );
+      this.setFilteredCategories(filteredFetchedCategories);
+    }
   }
 
-  get filteredCategoriesGroupings() {
-    const filteredCategoryIds = new Set();
+  async setFilterAndMode(newFilter, newMode) {
+    this.filter = newFilter;
+    this.mode = newMode;
 
-    const groupings = this.categoryGroupings.reduce((acc, categoryGrouping) => {
-      const filteredCategories = new Set();
+    if (!this.processing) {
+      this.processing = true;
 
-      const addCategory = (category) => {
-        if (this.#matchesFilter(category)) {
-          if (category.parentCategory?.parentCategory) {
-            filteredCategories.add(category.parentCategory.parentCategory);
-          }
-
-          if (category.parentCategory) {
-            filteredCategories.add(category.parentCategory);
-          }
+      try {
+        while (true) {
```

Comment:
I'm not sure if the `while` loop is necessary here. Could we move setting `processing` to `true` at the beginning of `searchCategories` and then setting it back to `false` when that async call is done?

Reply:
This is to serialize the requests. I'm not sending a request unless the previous one is complete, so then, if a request is required, it needs to happen immediately after the previous request.

Reply:
Ah, I see. I'm not sure we should wait for the previous request, in most cases, we usually cancel the previous request if there is a new. For example, that's what we do with search requests (see`search-menu.js`).

Reply:
The pattern in search-menu.js *does* have a race condition. There are two promises, the one for the request and the one that is constructed from that. The latter resolves in a microtask after the first promise resolves. So, it's possible to abort the request and still have the latter promise resolving after aborting.

Reply:
It's also difficult to cancel this request. It's ok if the request is for the search, but if the filter is blank, the request is for asyncFindByIds, which goes through a cache. I'd have to reference count the pending requests in the cache in order to cancel the correct http requests.

Reply:
Ok, I understand. Let's leave this as is.